### PR TITLE
8308151: [JVMCI] capture JVMCI exceptions in hs-err

### PR DESCRIPTION
**Pull request by dougxc**
When there is a pending exception after a JVMCI upcall into libjvmci, the VM calls the ExceptionDescribe JNI function to print the exception. Unfortunately, this output goes to "a system error-reporting channel" [1] which may not be tty. It also means the output is not in a hs-err log should the VM then exit with a fatal error. This has historically made it harder to triage libgraal bugs (i.e. the console output is usually required in addition to the hs-err crash log).

This PR addresses these shortcomings by printing the exception info to a string which is added to the JVMCI event log (for hs-err):
```
JVMCI Events (11 events):
...
Event: 0.274 Thread 0x0000000146819210 compiler.jvmci.TestUncaughtErrorInCompileMethod$CompilerCreationError
Event: 0.274 Thread 0x0000000146819210  at compiler.jvmci.TestUncaughtErrorInCompileMethod$1.createCompiler(TestUncaughtErrorInCompileMethod.java:147)
Event: 0.274 Thread 0x0000000146819210  at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.getCompiler(HotSpotJVMCIRuntime.java:829)
Event: 0.274 Thread 0x0000000146819210  at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.compileMethod(HotSpotJVMCIRuntime.java:943)
```

It is also be used to enhance the `-XX:+PrintCompilation` message issued for a failed compilation:
```
COMPILE SKIPPED: uncaught exception in call_HotSpotJVMCIRuntime_compileMethod [compiler.jvmci.TestUncaughtErrorInCompileMethod$CompilerCreationError]
```

[1] https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#exceptiondescribe

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308151](https://bugs.openjdk.org/browse/JDK-8308151): [JVMCI] capture JVMCI exceptions in hs-err

### Reviewers
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [b1854a6a](https://git.openjdk.org/jdk/pull/14000/files/b1854a6acb9e2bab9b871d112a0da9886f8a9c21)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14000/head:pull/14000` \
`$ git checkout pull/14000`

Update a local copy of the PR: \
`$ git checkout pull/14000` \
`$ git pull https://git.openjdk.org/jdk.git pull/14000/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14000`

View PR using the GUI difftool: \
`$ git pr show -t 14000`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14000.diff">https://git.openjdk.org/jdk/pull/14000.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14000#issuecomment-1549287673)

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 1]
```diff
@@ -2041,6 +2043,11 @@ void JVMCIRuntime::compile_method(JVMCIEnv* JVMCIENV, JVMCICompiler* compiler, c
 
   JVMCIObject result_object = JVMCIENV->call_HotSpotJVMCIRuntime_compileMethod(receiver, jvmci_method, entry_bci,
                                                                      (jlong) compile_state, compile_state->task()->compile_id());
+#ifdef ASSERT
+  if (JVMCIENV->has_pending_exception() && JVMCICompileMethodExceptionIsFatal) {
```
[the remark this answers is not in the document]
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp:2047]: Why can't you use a JVMCI property for this?  You get a chance to see them when copying them to Graal.

**Review by tkrodriguez — APPROVED** [review]

File: src/hotspot/share/jvmci/jvmciEnv.cpp [hunk 2]
```diff
@@ -310,25 +310,128 @@ void JVMCIEnv::init(JavaThread* thread, bool is_hotspot, const char* file, int l
...
+        *eol = '\0';
+        if (line < max_lines) {
+          JVMCI_event_1("%s", cursor);
+        }
+        cursor = eol + 1;
+      }
+      line++;
+    }
+    if (line >= max_lines) {
+      JVMCI_event_1("[elided %d more stack trace lines]", line - max_lines);
```
[the remark this answers is not in the document]
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciEnv.cpp:342]: Indeed:
```
Event: 0.237 Thread 0x000000011e019e10 compiler.jvmci.TestUncaughtErrorInCompileMethod$CompilerCreationError
Event: 0.237 Thread 0x000000011e019e10 	at compiler.jvmci.TestUncaughtErrorInCompileMethod$1.createCompiler(TestUncaughtErrorInCompileMethod.java:161) [elided 2 more stack trace lines]
```

File: src/hotspot/share/jvmci/jvmciRuntime.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp:2047]: The copying is only done when using libgraal. I'd like to have this test run in a JDK without libgraal.

[the remark this answers is not in the document]
Comment (dougxc) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp]: I see that I can simply use system properties after all: https://github.com/openjdk/jdk/pull/14000/commits/90f4346b3c8737fd0fee25a7ed0c32a1bd506c88

[the remark this answers is not in the document]
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp]: Yes that looks good.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 3]
```diff
@@ -2041,6 +2044,13 @@ void JVMCIRuntime::compile_method(JVMCIEnv* JVMCIENV, JVMCICompiler* compiler, c
 
   JVMCIObject result_object = JVMCIENV->call_HotSpotJVMCIRuntime_compileMethod(receiver, jvmci_method, entry_bci,
                                                                      (jlong) compile_state, compile_state->task()->compile_id());
+  if (JVMCIENV->has_pending_exception()) {
+    const char* val = Arguments::PropertyList_get_value(Arguments::system_properties(), "test.jvmci.compileMethodExceptionIsFatal");
```
Comment (dougxc) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:2048]: Note that this view on system properties is restricted to properties set at VM startup (e.g. on the command line) and will not see the result of calls to `System.setProperty()` made by an application.

**Comment by tkrodriguez** [comment]
The latest version looks ok to me.

**Pull request closed without merging**